docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+
+A series of changes have been landed to improve some of the fundamentals of Beaker in preparation for production launch:
+
+- [Various improvements](https://github.com/algorand-devrel/beaker/pull/142) - Improved test coverage, improved dev experience (setup + ongoing) via Poetry, improvements to the code quality setup (linting, automatic formatting, typing), allowed Windows development on Beaker itself, significantly improved CI/CD pipeline speed, removing the examples directory and tests from being distributed wit hthe PyPi package
+- [Typing improvements](https://github.com/algorand-devrel/beaker/pull/147)
+- [Removed inline imports](https://github.com/algorand-devrel/beaker/pull/148)
+- [Removed dead code](https://github.com/algorand-devrel/beaker/pull/149/files)
+- [Added automated release management and versioning](https://github.com/algorand-devrel/beaker/pull/161)
+
+In addition, there are a remaining set of more major (breaking) changes that are recommended. The recommendations are split into 2 categories, recommendations for immediate improvement (i.e. included in v1.0) and future suggestions that can be addressed post v1.0 launch.
+
+The recommended additional areas for immediate improvement are:
+
+- **Replace the class-based structure with an instance based one** - remove some areas of potential surprise for developers and simplify the Beaker codebase by moving to a composable instance-based structure rather than a static class-based structure
+- **Defer PyTEAL compilation** - improve flexibility and future contract output stability by deferring PyTEAL compilation (i.e. Beaker -> TEAL transpilation) to not happen when the Beaker contract is initialised
+- **Renamings** - There are some clear parameters that make sense to rename for various reasons
+- **Key decorator improvements** - Refactor some of the Beaker decorators to fix some bugs and improve user experience
+- **Beaker state refactor** - Refactor of the Beaker state interfaces to improve user extensibility and significantly simplify the Beaker codebase to improve maintainability
+
+The recommended areas for future improvement are:
+
+- Typed client generation from app spec to improve deploy-time and run-time dev experience
+- `Tmpl` values in app spec so you can have type-safe deployment clients that substitute any template values reliably at contract deploy time
+- Refactor storage types (blob, reserved, etc.) to allow use of in-built Python types and operators (terser, more intuitive)
+- Box storage implementation improved to match local/global behaviour and also automatically delete itself on contract deletion
+- Composable and stackable authorization and `@authorize` as a standalone decorator
+- PyTEAL typings to be improved to support types beyond `Expr` where a more explicit type can be specified (improves typing and extensibility)
```

Comment:
Can we get a list of the typings?

Reply:
So with PyTEAL's current architecture, `Expr` is effectively completely dynamically typed. The type is encoded into `Expr.type_of()`. If those types could be encoded in the Python type system instead (e.g. `UInt64Expr`, `BytesExpr`), it would make working with PyTEAL a lot easier and more intuitive.